{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "######################################################################\n",
    "#\n",
    "# Name: eQTL Pre-Processing\n",
    "# Author: Gabe Foster\n",
    "# Date: 5/15/2020\n",
    "# Purpose: This script performs the initial preprocessing of the \n",
    "# NF54HT-GFP-luc x NHP4026 transcriptional cross section\n",
    "#\n",
    "######################################################################\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "######################################################################\n",
    "#\n",
    "# Read block- let's get our file, set our variables, and perform the\n",
    "# initial filtering\n",
    "#\n",
    "######################################################################\n",
    "\n",
    "rawcounts = pd.read_csv('GENE2.count',\n",
    "                       sep = '\\t')\n",
    "\n",
    "# Setting fragment length- we can't see genes smaller than this and\n",
    "# calculating TPM correctly requires it. Tx ran this as 250bp PE,\n",
    "# and I don't have the original data so we're ballparking at 250.\n",
    "\n",
    "fraglength = 250\n",
    "\n",
    "# Remove genes shorter than frag length\n",
    "\n",
    "rawcounts_filt = rawcounts[rawcounts['Length'] > fraglength].T\n",
    "\n",
    "\n",
    "# We have a mess of strain name issues to fix here, as well\n",
    "\n",
    "rawcounts_filt.index = rawcounts_filt.index.str.replace('\\/','', regex = True)\n",
    "rawcounts_filt.index = rawcounts_filt.index.str.replace('ND5A5', 'AC075', regex = True)\n",
    "rawcounts_filt.index = rawcounts_filt.index.str.replace('ND6G8', 'AC125', regex = True)\n",
    "rawcounts_filt.index = rawcounts_filt.index.str.replace('N1', '', regex = True)\n",
    "rawcounts_filt.index = rawcounts_filt.index.str.replace('\\\\.', '', regex = True)\n",
    "rawcounts_filt.index = rawcounts_filt.index.str.replace('_4026', '_NHP4026', regex = True)\n",
    "rawcounts_filt.index = rawcounts_filt.index.str.replace('^4026', 'NHP4026', regex = True)\n",
    "rawcounts_filt.index = rawcounts_filt.index.str.replace('2H9', 'AC030', regex = True)\n",
    "rawcounts_filt.index = rawcounts_filt.index.str.replace('6E5', 'AC033', regex = True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "######################################################################\n",
    "#\n",
    "# Initial sample QC- drop samples with < 3000 genes represented, zero\n",
    "# any sample/gene reads < 5, and pick rep with most counts\n",
    "#\n",
    "######################################################################\n",
    "\n",
    "# Keep only samples with counts in > 3000 genes, and sum counts for\n",
    "# each sample along the way\n",
    "good_samples = []\n",
    "sample_sum = []\n",
    "for i in range(6,len(rawcounts_filt.index)):\n",
    "    genes_rep = np.sum(rawcounts_filt.iloc[i,:] > 0)\n",
    "    if genes_rep > 3000:\n",
    "        good_samples.append(rawcounts_filt.index[i])\n",
    "        sample_sum.append(np.sum(rawcounts_filt.iloc[i,:]))\n",
    "        \n",
    "# Build summary frame, iterate over samples by strain/hpi and choose \n",
    "# the sample with the most reads to keep\n",
    "\n",
    "count_summary = {'Sample Name':good_samples,\n",
    "                 'Total Counts':sample_sum}\n",
    "count_summary = pd.DataFrame(count_summary)\n",
    "\n",
    "# This is a hot mess of regular expression that just converts the entire\n",
    "# sample name to strain_##, where ## is the sampling timepoint. The formatting\n",
    "# is very inconsistent throughout, so a mess of replacements need to be carefully\n",
    "# made. This mess of substitutions makes them.\n",
    "\n",
    "count_summary['Strain and Time'] = count_summary['Sample Name'].str.replace('^GF_PL[\\d]+[a,b,c]{0,1}_', '',regex = True)\n",
    "count_summary['Strain and Time'] = count_summary['Strain and Time'].str.replace('[A,B,C]_', '', regex = True)\n",
    "count_summary['Strain and Time'] = count_summary['Strain and Time'].str.replace('_[d]+$', '', regex = True)\n",
    "count_summary['Strain and Time'] = count_summary['Strain and Time'].str.replace('_S.*', '', regex = True)\n",
    "count_summary['Strain and Time'] = count_summary['Strain and Time'].str.replace('_[0-9]{3,4}$', '', regex = True)\n",
    "count_summary['Strain and Time'] = count_summary['Strain and Time'].str.replace('hpi', '', regex = True)\n",
    "count_summary['Strain and Time'] = count_summary['Strain and Time'].str.replace('_T', '_', regex = True)\n",
    "\n",
    "# Believe it or not, the GTEx consortium handled reps across batches and such by...\n",
    "# taking the replicate with the most counts. That's easy, let's do that.\n",
    "\n",
    "best_samples = []\n",
    "for sample in count_summary['Strain and Time'].unique():\n",
    "    subframe = count_summary[count_summary['Strain and Time'] == sample]\n",
    "    best_samples.append(count_summary.iloc[subframe['Total Counts'].idxmax(),:])\n",
    "    \n",
    "# Now I'll build a frame with just the best samples in it\n",
    "    \n",
    "best_samples = pd.DataFrame(best_samples)\n",
    "best_samplenames = ['Geneid', 'Chr', 'Start', 'End', 'Strand', 'Length']\n",
    "best_samplenames.extend(list(best_samples['Sample Name']))\n",
    "curated_counts = rawcounts_filt[rawcounts_filt.index.isin(best_samplenames)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 285,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AC033</th>\n",
       "      <th>AC004</th>\n",
       "      <th>AC006</th>\n",
       "      <th>AC007</th>\n",
       "      <th>AC008</th>\n",
       "      <th>AC025</th>\n",
       "      <th>AC027</th>\n",
       "      <th>AC028</th>\n",
       "      <th>AC034</th>\n",
       "      <th>AC049</th>\n",
       "      <th>...</th>\n",
       "      <th>AC118</th>\n",
       "      <th>AC081</th>\n",
       "      <th>AC038</th>\n",
       "      <th>AC082</th>\n",
       "      <th>AC030</th>\n",
       "      <th>AC032</th>\n",
       "      <th>AC074</th>\n",
       "      <th>AC130</th>\n",
       "      <th>NF54gfp</th>\n",
       "      <th>AC075</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>malmito_rna_16:rRNA</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>malmito_rna_20:rRNA</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>malmito_rna_9:rRNA</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>malmito_rna_17:rRNA</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>malmito_rna_LSUC:rRNA</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF3D7_1479700</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF3D7_1479800</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF3D7_1479900</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF3D7_1480000</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF3D7_1480100</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5772 rows × 53 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       AC033  AC004  AC006  AC007  AC008  AC025  AC027  AC028  \\\n",
       "Geneid                                                                          \n",
       "malmito_rna_16:rRNA        0      0      0      0      0      0      1      0   \n",
       "malmito_rna_20:rRNA        0      0      0      0      0      0      1      0   \n",
       "malmito_rna_9:rRNA         0      0      0      0      0      0      0      0   \n",
       "malmito_rna_17:rRNA        0      0      0      0      0      0      1      0   \n",
       "malmito_rna_LSUC:rRNA      0      0      0      0      0      0      0      0   \n",
       "...                      ...    ...    ...    ...    ...    ...    ...    ...   \n",
       "PF3D7_1479700              0      0      0      1      4      0      0      0   \n",
       "PF3D7_1479800              0      0      0      0      0      0      0      0   \n",
       "PF3D7_1479900              0      0      0      0      0      0      1      0   \n",
       "PF3D7_1480000              1      0      0      0      0      0      0      0   \n",
       "PF3D7_1480100              0      0      0      0      0      0      0      0   \n",
       "\n",
       "                       AC034  AC049  ...  AC118  AC081  AC038  AC082  AC030  \\\n",
       "Geneid                               ...                                      \n",
       "malmito_rna_16:rRNA        0      0  ...      0      0      0      0      0   \n",
       "malmito_rna_20:rRNA        0      0  ...      0      1      0      0      0   \n",
       "malmito_rna_9:rRNA         0      0  ...      0      0      0      0      0   \n",
       "malmito_rna_17:rRNA        0      0  ...      0      0      0      1      0   \n",
       "malmito_rna_LSUC:rRNA      2      0  ...      3      0      0      0      0   \n",
       "...                      ...    ...  ...    ...    ...    ...    ...    ...   \n",
       "PF3D7_1479700              3      0  ...      0      0      1      0      2   \n",
       "PF3D7_1479800              1      0  ...      1      0      1      0      0   \n",
       "PF3D7_1479900              0      0  ...      0      0      0      1      0   \n",
       "PF3D7_1480000              1      0  ...      1      1      0      0      0   \n",
       "PF3D7_1480100              3      1  ...      0      1      0      0      0   \n",
       "\n",
       "                       AC032  AC074  AC130  NF54gfp  AC075  \n",
       "Geneid                                                      \n",
       "malmito_rna_16:rRNA        1      0      0        0      0  \n",
       "malmito_rna_20:rRNA        3      1      1        0      0  \n",
       "malmito_rna_9:rRNA         0      0      0        0      0  \n",
       "malmito_rna_17:rRNA        1      0      0        1      0  \n",
       "malmito_rna_LSUC:rRNA      4      1      0        1      0  \n",
       "...                      ...    ...    ...      ...    ...  \n",
       "PF3D7_1479700              0      1      0        1      0  \n",
       "PF3D7_1479800              0      1      0        1      0  \n",
       "PF3D7_1479900              0      0      0        0      0  \n",
       "PF3D7_1480000              1      2      1        1      0  \n",
       "PF3D7_1480100              1      0      0        0      0  \n",
       "\n",
       "[5772 rows x 53 columns]"
      ]
     },
     "execution_count": 285,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rawcounts = pd.read_csv('T4_counts.txt', \n",
    "                          sep = '\\t', index_col=0)\n",
    "rawcounts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sample_and_geneqc(expdata):\n",
    "    '''\n",
    "    Here, we'll remove poor samples and poor genes, and QC low read #s\n",
    "    Poor samples are defined as those with < 3000 genes with reads\n",
    "    Poor genes are those that appear in < 20% of samples\n",
    "    Low reads are any gene/sample reading < 5 counts; those are zeroed out\n",
    "    '''\n",
    "    \n",
    "    # Curate Samples\n",
    "    \n",
    "    genecounts = pd.Series(data = np.count_nonzero(expdata, axis = 1),\n",
    "                             index = expdata.index)\n",
    "    samplecounts = pd.Series(np.count_nonzero(expdata,axis = 0),\n",
    "                          index = expdata.columns)\n",
    "    \n",
    "    goodgenes = genecounts[genecounts/samplecounts.size > 0.2]\n",
    "    \n",
    "    goodsamples = samplecounts[samplecounts > 3000]\n",
    "    \n",
    "    allcur = expdata.loc[goodgenes.index, goodsamples.index]\n",
    "    \n",
    "    # Zero out counts < 5\n",
    "    \n",
    "    allcur = allcur.mask(allcur < 5, 0)\n",
    "    \n",
    "    return allcur"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 287,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AC033</th>\n",
       "      <th>AC004</th>\n",
       "      <th>AC006</th>\n",
       "      <th>AC007</th>\n",
       "      <th>AC008</th>\n",
       "      <th>AC025</th>\n",
       "      <th>AC027</th>\n",
       "      <th>AC028</th>\n",
       "      <th>AC034</th>\n",
       "      <th>AC049</th>\n",
       "      <th>...</th>\n",
       "      <th>AC118</th>\n",
       "      <th>AC081</th>\n",
       "      <th>AC038</th>\n",
       "      <th>AC082</th>\n",
       "      <th>AC030</th>\n",
       "      <th>AC032</th>\n",
       "      <th>AC074</th>\n",
       "      <th>AC130</th>\n",
       "      <th>NF54gfp</th>\n",
       "      <th>AC075</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>malmito_rna_20:rRNA</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>malmito_rna_17:rRNA</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>malmito_rna_LSUC:rRNA</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>malmito_rna_LSUG:rRNA</th>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>malmito_SSUB:rRNA</th>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF3D7_1479200</th>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>15</td>\n",
       "      <td>12</td>\n",
       "      <td>33</td>\n",
       "      <td>49</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>34</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "      <td>22</td>\n",
       "      <td>24</td>\n",
       "      <td>14</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF3D7_1479400</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF3D7_1479700</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF3D7_1479800</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF3D7_1480000</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5496 rows × 53 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       AC033  AC004  AC006  AC007  AC008  AC025  AC027  AC028  \\\n",
       "Geneid                                                                          \n",
       "malmito_rna_20:rRNA        0      0      0      0      0      0      0      0   \n",
       "malmito_rna_17:rRNA        0      0      0      0      0      0      0      0   \n",
       "malmito_rna_LSUC:rRNA      0      0      0      0      0      0      0      0   \n",
       "malmito_rna_LSUG:rRNA      0      7      5      0      0      0      0      0   \n",
       "malmito_SSUB:rRNA          0      5      0      0      0      0      0      0   \n",
       "...                      ...    ...    ...    ...    ...    ...    ...    ...   \n",
       "PF3D7_1479200              0      5      0      5     15     12     33     49   \n",
       "PF3D7_1479400              0      0      0      0      0      0      0      0   \n",
       "PF3D7_1479700              0      0      0      0      0      0      0      0   \n",
       "PF3D7_1479800              0      0      0      0      0      0      0      0   \n",
       "PF3D7_1480000              0      0      0      0      0      0      0      0   \n",
       "\n",
       "                       AC034  AC049  ...  AC118  AC081  AC038  AC082  AC030  \\\n",
       "Geneid                               ...                                      \n",
       "malmito_rna_20:rRNA        0      0  ...      0      0      0      0      0   \n",
       "malmito_rna_17:rRNA        0      0  ...      0      0      0      0      0   \n",
       "malmito_rna_LSUC:rRNA      0      0  ...      0      0      0      0      0   \n",
       "malmito_rna_LSUG:rRNA      6      0  ...     15      0      0      6      5   \n",
       "malmito_SSUB:rRNA          0      0  ...      5      0      0      0      0   \n",
       "...                      ...    ...  ...    ...    ...    ...    ...    ...   \n",
       "PF3D7_1479200              0     28  ...      6      0      8     34      5   \n",
       "PF3D7_1479400              0      0  ...      0      0      0      0      0   \n",
       "PF3D7_1479700              0      0  ...      0      0      0      0      0   \n",
       "PF3D7_1479800              0      0  ...      0      0      0      0      0   \n",
       "PF3D7_1480000              0      0  ...      0      0      0      0      0   \n",
       "\n",
       "                       AC032  AC074  AC130  NF54gfp  AC075  \n",
       "Geneid                                                      \n",
       "malmito_rna_20:rRNA        0      0      0        0      0  \n",
       "malmito_rna_17:rRNA        0      0      0        0      0  \n",
       "malmito_rna_LSUC:rRNA      0      0      0        0      0  \n",
       "malmito_rna_LSUG:rRNA     12      0      0       10      0  \n",
       "malmito_SSUB:rRNA          9      0      0        0      0  \n",
       "...                      ...    ...    ...      ...    ...  \n",
       "PF3D7_1479200              9     22     24       14     21  \n",
       "PF3D7_1479400              0      0      0        0      0  \n",
       "PF3D7_1479700              0      0      0        0      0  \n",
       "PF3D7_1479800              0      0      0        0      0  \n",
       "PF3D7_1480000              0      0      0        0      0  \n",
       "\n",
       "[5496 rows x 53 columns]"
      ]
     },
     "execution_count": 287,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test = sample_and_geneqc(rawcounts)\n",
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 383,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tpm_norm(gctdata, probefile, fraglength = 250):\n",
    "    '''\n",
    "    TPM is (reads / effective gene length (in kb)) / (sample reads / 1e6)\n",
    "    The elaborate nonsense below just calculates this, it's a bit of a\n",
    "    mess but hey that's programming\n",
    "    '''\n",
    "    \n",
    "    # Read in probe metadata, subtract frag length,\n",
    "    # and convert to dict\n",
    "    \n",
    "    \n",
    "    probeinfo = pd.read_csv(probefile, usecols = [1,6])\n",
    "    probeinfo.index = probeinfo['Geneid']\n",
    "    probeinfo.drop(columns = 'Geneid', inplace = True)\n",
    "    probeinfo = probeinfo[probeinfo.index.isin(gctdata.index)]\n",
    "    probeinfo['Length'] = probeinfo['Length'].apply(lambda x: x-fraglength)\n",
    "        \n",
    "    # Build output frame from input frame (cheating), calc length and\n",
    "    # build lookup from samplename to total counts\n",
    "    \n",
    "    tpm = gctdata.copy()\n",
    "    curated_tpm = gctdata.copy()\n",
    "    \n",
    "    # Iterate over rows- if non sample rows, just copy data, if sample\n",
    "    # rows, calculate TPM\n",
    "    \n",
    "    # Note: it really doesn't like the way I did this and will throw\n",
    "    # warnings- it does work, it's fine, don't worry about it, if you\n",
    "    # are smarter than I am go ahead and rewrite it\n",
    "    \n",
    "    for (col, data) in gctdata.iteritems():\n",
    "        \n",
    "        numreads = data.sum()\n",
    "        \n",
    "        tempframe = pd.concat([probeinfo, data], axis = 1)\n",
    "\n",
    "        tempframe['rpkb'] = tempframe.iloc[:,1].divide(tempframe['Length'].divide(1000))\n",
    "        tempframe['tpm'] = (tempframe['rpkb'] / data.sum()) * 1e6\n",
    "        tempframe['tpm'] = tempframe['tpm'].clip(lower = 0)\n",
    "        curated_tpm.loc[:,col] = tempframe['tpm']\n",
    "    return curated_tpm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 384,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AC033</th>\n",
       "      <th>AC004</th>\n",
       "      <th>AC006</th>\n",
       "      <th>AC007</th>\n",
       "      <th>AC008</th>\n",
       "      <th>AC025</th>\n",
       "      <th>AC027</th>\n",
       "      <th>AC028</th>\n",
       "      <th>AC034</th>\n",
       "      <th>AC049</th>\n",
       "      <th>...</th>\n",
       "      <th>AC118</th>\n",
       "      <th>AC081</th>\n",
       "      <th>AC038</th>\n",
       "      <th>AC082</th>\n",
       "      <th>AC030</th>\n",
       "      <th>AC032</th>\n",
       "      <th>AC074</th>\n",
       "      <th>AC130</th>\n",
       "      <th>NF54gfp</th>\n",
       "      <th>AC075</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>malmito_rna_20:rRNA</th>\n",
       "      <td>-0.0</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.0</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.00000</td>\n",
       "      <td>-0.0</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.0</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>malmito_rna_17:rRNA</th>\n",
       "      <td>-0.0</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.0</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.00000</td>\n",
       "      <td>-0.0</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.0</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>malmito_rna_LSUC:rRNA</th>\n",
       "      <td>-0.0</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.0</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.00000</td>\n",
       "      <td>-0.0</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.0</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>malmito_rna_LSUG:rRNA</th>\n",
       "      <td>-0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.0</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>malmito_SSUB:rRNA</th>\n",
       "      <td>-0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.0</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.00000</td>\n",
       "      <td>-0.0</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.0</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF3D7_1479200</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.767031</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.115203</td>\n",
       "      <td>24.895135</td>\n",
       "      <td>18.831</td>\n",
       "      <td>33.382695</td>\n",
       "      <td>86.70573</td>\n",
       "      <td>0.0</td>\n",
       "      <td>65.149443</td>\n",
       "      <td>...</td>\n",
       "      <td>8.726986</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.155266</td>\n",
       "      <td>28.595928</td>\n",
       "      <td>4.282387</td>\n",
       "      <td>8.378456</td>\n",
       "      <td>25.614292</td>\n",
       "      <td>19.650779</td>\n",
       "      <td>14.324915</td>\n",
       "      <td>21.31184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF3D7_1479400</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF3D7_1479700</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF3D7_1479800</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF3D7_1480000</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5496 rows × 53 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       AC033     AC004  AC006     AC007      AC008   AC025  \\\n",
       "Geneid                                                                       \n",
       "malmito_rna_20:rRNA     -0.0 -0.000000   -0.0 -0.000000  -0.000000  -0.000   \n",
       "malmito_rna_17:rRNA     -0.0 -0.000000   -0.0 -0.000000  -0.000000  -0.000   \n",
       "malmito_rna_LSUC:rRNA   -0.0 -0.000000   -0.0 -0.000000  -0.000000  -0.000   \n",
       "malmito_rna_LSUG:rRNA   -0.0  0.000000    0.0 -0.000000  -0.000000  -0.000   \n",
       "malmito_SSUB:rRNA       -0.0  0.000000   -0.0 -0.000000  -0.000000  -0.000   \n",
       "...                      ...       ...    ...       ...        ...     ...   \n",
       "PF3D7_1479200            0.0  7.767031    0.0  9.115203  24.895135  18.831   \n",
       "PF3D7_1479400            0.0  0.000000    0.0  0.000000   0.000000   0.000   \n",
       "PF3D7_1479700            0.0  0.000000    0.0  0.000000   0.000000   0.000   \n",
       "PF3D7_1479800            0.0  0.000000    0.0  0.000000   0.000000   0.000   \n",
       "PF3D7_1480000            0.0  0.000000    0.0  0.000000   0.000000   0.000   \n",
       "\n",
       "                           AC027     AC028  AC034      AC049  ...     AC118  \\\n",
       "Geneid                                                        ...             \n",
       "malmito_rna_20:rRNA    -0.000000  -0.00000   -0.0  -0.000000  ... -0.000000   \n",
       "malmito_rna_17:rRNA    -0.000000  -0.00000   -0.0  -0.000000  ... -0.000000   \n",
       "malmito_rna_LSUC:rRNA  -0.000000  -0.00000   -0.0  -0.000000  ... -0.000000   \n",
       "malmito_rna_LSUG:rRNA  -0.000000  -0.00000    0.0  -0.000000  ...  0.000000   \n",
       "malmito_SSUB:rRNA      -0.000000  -0.00000   -0.0  -0.000000  ...  0.000000   \n",
       "...                          ...       ...    ...        ...  ...       ...   \n",
       "PF3D7_1479200          33.382695  86.70573    0.0  65.149443  ...  8.726986   \n",
       "PF3D7_1479400           0.000000   0.00000    0.0   0.000000  ...  0.000000   \n",
       "PF3D7_1479700           0.000000   0.00000    0.0   0.000000  ...  0.000000   \n",
       "PF3D7_1479800           0.000000   0.00000    0.0   0.000000  ...  0.000000   \n",
       "PF3D7_1480000           0.000000   0.00000    0.0   0.000000  ...  0.000000   \n",
       "\n",
       "                       AC081      AC038      AC082     AC030     AC032  \\\n",
       "Geneid                                                                   \n",
       "malmito_rna_20:rRNA     -0.0  -0.000000  -0.000000 -0.000000 -0.000000   \n",
       "malmito_rna_17:rRNA     -0.0  -0.000000  -0.000000 -0.000000 -0.000000   \n",
       "malmito_rna_LSUC:rRNA   -0.0  -0.000000  -0.000000 -0.000000 -0.000000   \n",
       "malmito_rna_LSUG:rRNA   -0.0  -0.000000   0.000000  0.000000  0.000000   \n",
       "malmito_SSUB:rRNA       -0.0  -0.000000  -0.000000 -0.000000  0.000000   \n",
       "...                      ...        ...        ...       ...       ...   \n",
       "PF3D7_1479200            0.0  11.155266  28.595928  4.282387  8.378456   \n",
       "PF3D7_1479400            0.0   0.000000   0.000000  0.000000  0.000000   \n",
       "PF3D7_1479700            0.0   0.000000   0.000000  0.000000  0.000000   \n",
       "PF3D7_1479800            0.0   0.000000   0.000000  0.000000  0.000000   \n",
       "PF3D7_1480000            0.0   0.000000   0.000000  0.000000  0.000000   \n",
       "\n",
       "                           AC074      AC130    NF54gfp     AC075  \n",
       "Geneid                                                            \n",
       "malmito_rna_20:rRNA    -0.000000  -0.000000  -0.000000  -0.00000  \n",
       "malmito_rna_17:rRNA    -0.000000  -0.000000  -0.000000  -0.00000  \n",
       "malmito_rna_LSUC:rRNA  -0.000000  -0.000000  -0.000000  -0.00000  \n",
       "malmito_rna_LSUG:rRNA  -0.000000  -0.000000   0.000000  -0.00000  \n",
       "malmito_SSUB:rRNA      -0.000000  -0.000000  -0.000000  -0.00000  \n",
       "...                          ...        ...        ...       ...  \n",
       "PF3D7_1479200          25.614292  19.650779  14.324915  21.31184  \n",
       "PF3D7_1479400           0.000000   0.000000   0.000000   0.00000  \n",
       "PF3D7_1479700           0.000000   0.000000   0.000000   0.00000  \n",
       "PF3D7_1479800           0.000000   0.000000   0.000000   0.00000  \n",
       "PF3D7_1480000           0.000000   0.000000   0.000000   0.00000  \n",
       "\n",
       "[5496 rows x 53 columns]"
      ]
     },
     "execution_count": 384,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test2 = tpm_norm(test, 'GENE2_probeinfo.csv', 250)\n",
    "test2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 307,
   "metadata": {},
   "outputs": [],
   "source": [
    "def testfunc(expdata, probedata, fraglength = 250):\n",
    "    return expdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AC033</th>\n",
       "      <th>AC004</th>\n",
       "      <th>AC006</th>\n",
       "      <th>AC007</th>\n",
       "      <th>AC008</th>\n",
       "      <th>AC025</th>\n",
       "      <th>AC027</th>\n",
       "      <th>AC028</th>\n",
       "      <th>AC034</th>\n",
       "      <th>AC049</th>\n",
       "      <th>...</th>\n",
       "      <th>AC118</th>\n",
       "      <th>AC081</th>\n",
       "      <th>AC038</th>\n",
       "      <th>AC082</th>\n",
       "      <th>AC030</th>\n",
       "      <th>AC032</th>\n",
       "      <th>AC074</th>\n",
       "      <th>AC130</th>\n",
       "      <th>NF54gfp</th>\n",
       "      <th>AC075</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>malmito_rna_20:rRNA</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>malmito_rna_17:rRNA</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>malmito_rna_LSUC:rRNA</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>malmito_rna_LSUG:rRNA</th>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>malmito_SSUB:rRNA</th>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF3D7_1479200</th>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>15</td>\n",
       "      <td>12</td>\n",
       "      <td>33</td>\n",
       "      <td>49</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>34</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "      <td>22</td>\n",
       "      <td>24</td>\n",
       "      <td>14</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF3D7_1479400</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF3D7_1479700</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF3D7_1479800</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF3D7_1480000</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5496 rows × 53 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       AC033  AC004  AC006  AC007  AC008  AC025  AC027  AC028  \\\n",
       "Geneid                                                                          \n",
       "malmito_rna_20:rRNA        0      0      0      0      0      0      0      0   \n",
       "malmito_rna_17:rRNA        0      0      0      0      0      0      0      0   \n",
       "malmito_rna_LSUC:rRNA      0      0      0      0      0      0      0      0   \n",
       "malmito_rna_LSUG:rRNA      0      7      5      0      0      0      0      0   \n",
       "malmito_SSUB:rRNA          0      5      0      0      0      0      0      0   \n",
       "...                      ...    ...    ...    ...    ...    ...    ...    ...   \n",
       "PF3D7_1479200              0      5      0      5     15     12     33     49   \n",
       "PF3D7_1479400              0      0      0      0      0      0      0      0   \n",
       "PF3D7_1479700              0      0      0      0      0      0      0      0   \n",
       "PF3D7_1479800              0      0      0      0      0      0      0      0   \n",
       "PF3D7_1480000              0      0      0      0      0      0      0      0   \n",
       "\n",
       "                       AC034  AC049  ...  AC118  AC081  AC038  AC082  AC030  \\\n",
       "Geneid                               ...                                      \n",
       "malmito_rna_20:rRNA        0      0  ...      0      0      0      0      0   \n",
       "malmito_rna_17:rRNA        0      0  ...      0      0      0      0      0   \n",
       "malmito_rna_LSUC:rRNA      0      0  ...      0      0      0      0      0   \n",
       "malmito_rna_LSUG:rRNA      6      0  ...     15      0      0      6      5   \n",
       "malmito_SSUB:rRNA          0      0  ...      5      0      0      0      0   \n",
       "...                      ...    ...  ...    ...    ...    ...    ...    ...   \n",
       "PF3D7_1479200              0     28  ...      6      0      8     34      5   \n",
       "PF3D7_1479400              0      0  ...      0      0      0      0      0   \n",
       "PF3D7_1479700              0      0  ...      0      0      0      0      0   \n",
       "PF3D7_1479800              0      0  ...      0      0      0      0      0   \n",
       "PF3D7_1480000              0      0  ...      0      0      0      0      0   \n",
       "\n",
       "                       AC032  AC074  AC130  NF54gfp  AC075  \n",
       "Geneid                                                      \n",
       "malmito_rna_20:rRNA        0      0      0        0      0  \n",
       "malmito_rna_17:rRNA        0      0      0        0      0  \n",
       "malmito_rna_LSUC:rRNA      0      0      0        0      0  \n",
       "malmito_rna_LSUG:rRNA     12      0      0       10      0  \n",
       "malmito_SSUB:rRNA          9      0      0        0      0  \n",
       "...                      ...    ...    ...      ...    ...  \n",
       "PF3D7_1479200              9     22     24       14     21  \n",
       "PF3D7_1479400              0      0      0        0      0  \n",
       "PF3D7_1479700              0      0      0        0      0  \n",
       "PF3D7_1479800              0      0      0        0      0  \n",
       "PF3D7_1480000              0      0      0        0      0  \n",
       "\n",
       "[5496 rows x 53 columns]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test2 = testfunc(test, 'GENE2_probeinfo.csv')\n",
    "test2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'curated_counts' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-197-b5d7912d81f0>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     11\u001b[0m \u001b[1;31m# Build output frame from input frame (cheating), calc length and\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     12\u001b[0m \u001b[1;31m# build lookup from samplename to total counts\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 13\u001b[1;33m \u001b[0mcurated_counts\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcolumns\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mcurated_counts\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'Geneid'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     14\u001b[0m \u001b[0mcurated_tpm\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mcurated_counts\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     15\u001b[0m \u001b[0mlength\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mcurated_counts\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m5\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m-\u001b[0m \u001b[0mfraglength\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'curated_counts' is not defined"
     ]
    }
   ],
   "source": [
    "######################################################################\n",
    "#\n",
    "# Time to calculate TPM\n",
    "#\n",
    "######################################################################\n",
    "\n",
    "# TPM is (reads / effective gene length (in kb)) / (sample reads / 1e6)\n",
    "# The elaborate nonsense below just calculates this, it's a bit of a\n",
    "# mess but hey that's programming\n",
    "\n",
    "# Build output frame from input frame (cheating), calc length and\n",
    "# build lookup from samplename to total counts\n",
    "curated_counts.columns = curated_counts.loc['Geneid',:]\n",
    "curated_tpm = curated_counts\n",
    "length = curated_counts.iloc[5,:] - fraglength\n",
    "sample_counts = pd.Series(list(best_samples['Total Counts']),\n",
    "                          index = list(best_samples['Sample Name'])).to_dict()\n",
    "\n",
    "# Iterate over rows- if non sample rows, just copy data, if sample\n",
    "# rows, calculate TPM\n",
    "\n",
    "# Note: it really doesn't like the way I did this and will throw\n",
    "# warnings- it does work, it's fine, don't worry about it, if you\n",
    "# are smarter than I am go ahead and rewrite it\n",
    "\n",
    "for row, data in curated_counts.iterrows():\n",
    "    if row in ['Geneid','Chr', 'Start', 'End', 'Strand', 'Length']:\n",
    "        curated_tpm.loc[row,:] = data\n",
    "    else:\n",
    "        rpkb = data / (length/1000)\n",
    "        tpm = (rpkb / sample_counts[row]) * 1e6\n",
    "        curated_tpm.loc[row,:] = tpm\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "######################################################################\n",
    "#\n",
    "# Subset block\n",
    "# I have 3 time points and I want both the count data and the TPM\n",
    "# for all 3, only on the samples that passed QC. Let's do it.\n",
    "#\n",
    "######################################################################\n",
    "\n",
    "T4_samples = best_samples[best_samples['Strain and Time'].str.contains('_4$', regex = True)]\n",
    "T30_samples = best_samples[best_samples['Strain and Time'].str.contains('_30$', regex = True)]\n",
    "T44_samples = best_samples[best_samples['Strain and Time'].str.contains('_44$', regex = True)]\n",
    "\n",
    "\n",
    "\n",
    "# Let's pull out the count data for each time point\n",
    "# Now that we're normalized we can minimize this to a pure count matrix\n",
    "\n",
    "rawcounts_filt.columns = rawcounts_filt.loc['Geneid',:]\n",
    "filler = np.repeat('filler', len(rawcounts_filt.columns))\n",
    "\n",
    "T4_counts = rawcounts_filt[rawcounts_filt.index.isin(T4_samples['Sample Name'])].T\n",
    "T4_counts.name = 'T4_counts'\n",
    "T4_counts.insert(loc = 0,column = 'Filler',value = filler)\n",
    "T30_counts = rawcounts_filt[rawcounts_filt.index.isin(T30_samples['Sample Name'])].T\n",
    "T30_counts.name = 'T30_counts'\n",
    "T30_counts.insert(loc = 0,column = 'Filler',value = filler)\n",
    "T44_counts = rawcounts_filt[rawcounts_filt.index.isin(T44_samples['Sample Name'])].T\n",
    "T44_counts.name = 'T44_counts'\n",
    "T44_counts.insert(loc = 0,column = 'Filler',value = filler)\n",
    "\n",
    "# Alright, same thing but for TPM\n",
    "\n",
    "T4_tpm = curated_tpm[curated_tpm.index.isin(T4_samples['Sample Name'])].T\n",
    "T4_tpm.name = 'T4_tpm'\n",
    "T4_tpm.insert(loc = 0,column = 'Filler',value = filler)\n",
    "T30_tpm = curated_tpm[curated_tpm.index.isin(T30_samples['Sample Name'])].T\n",
    "T30_tpm.name = 'T30_tpm'\n",
    "T30_tpm.insert(loc = 0,column = 'Filler',value = filler)\n",
    "T44_tpm = curated_tpm[curated_tpm.index.isin(T44_samples['Sample Name'])].T\n",
    "T44_tpm.name = 'T44_tpm'\n",
    "T44_tpm.insert(loc = 0,column = 'Filler',value = filler)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "######################################################################\n",
    "#\n",
    "# Writing all this to a .gct file is a hassle, let us embark on this\n",
    "# hassle here\n",
    "#\n",
    "######################################################################\n",
    "\n",
    "# Here's a function that will write the above data frames to acceptable\n",
    "# gct format\n",
    "\n",
    "def to_gct(file, expdata):\n",
    "    mrow = 0\n",
    "    mcol = 0\n",
    "    row = len(expdata.index)\n",
    "    col = len(expdata.columns)\n",
    "    \n",
    "    outfile = open(file, 'w')\n",
    "    outfile.write(f'1.0\\n')\n",
    "    outfile.write(f'{row}\\t{col}\\n')\n",
    "    outfile.close()\n",
    "    expdata.to_csv(file,\n",
    "                   mode = 'a',\n",
    "                   header = True,\n",
    "                   sep = '\\t',\n",
    "                   index_label = 'Gene ID',\n",
    "                   encoding='utf-8')\n",
    "    \n",
    "\n",
    "    \n",
    "# Build a list of the files we need to write out\n",
    "    \n",
    "files_to_write = [T4_counts,\n",
    "                  T30_counts,\n",
    "                  T44_counts,\n",
    "                  T4_tpm,\n",
    "                  T30_tpm, \n",
    "                  T44_tpm]\n",
    "\n",
    "# So, we need to think ahead a bit- we need to strip down sample\n",
    "# names to just strain so that the name are consistent across both\n",
    "# the gct files and the vcf file for future use. So, this loop will\n",
    "# take each frame, use a bunch of replacement regexes to strip the \n",
    "# column names to just strain ID, and write to gct.\n",
    "\n",
    "for edata in files_to_write:\n",
    "    edata.columns = edata.columns.str.replace('^GF[\\d]*_', '', regex = True)\n",
    "    edata.columns = edata.columns.str.replace('PL[\\da-z]*_', '', regex = True)\n",
    "    edata.columns = edata.columns.str.replace('[A-C]{1}_', '', regex = True)\n",
    "    edata.columns = edata.columns.str.partition('_').to_frame().iloc[:,0]\n",
    "    \n",
    "    to_gct(file = f'{edata.name}.gct',\n",
    "          expdata = edata)\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 386,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_covariates(metadata):\n",
    "    '''\n",
    "    Our samples were run across different plates; this\n",
    "    is the only real place I can account for technical\n",
    "    variation. Our sampling batches are confounded with\n",
    "    stage, so I am forced to rely on CRC and PEER to\n",
    "    pull anything in there out. To run PEER, I need\n",
    "    one-hot encoded plate data, so I do that here.\n",
    "    '''\n",
    "    encoding = pd.get_dummies(metadata['PlateID'])\n",
    "    metadata = metadata.merge(encoding,\n",
    "                              left_index = True, \n",
    "                              right_index = True)\n",
    "    metadata.drop(columns = 'PlateID',\n",
    "                 inplace = True)\n",
    "    \n",
    "    return metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 387,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SampleID</th>\n",
       "      <th>PL01</th>\n",
       "      <th>PL02</th>\n",
       "      <th>PL03</th>\n",
       "      <th>PL04</th>\n",
       "      <th>PL05</th>\n",
       "      <th>PL05a</th>\n",
       "      <th>PL05b</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2H9_T44_S98_L00</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2H9_T4_S96_L00</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4026_T30_S103_L00</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4026_T44_S104_L00</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4026_T4_S102_L00</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>443</th>\n",
       "      <td>ND5A5_T44_S112_L00</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444</th>\n",
       "      <td>ND5A5_T4_S110_L00</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>445</th>\n",
       "      <td>ND6G8_T30_S114_L00</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>446</th>\n",
       "      <td>ND6G8_T44_S115_L00</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447</th>\n",
       "      <td>ND6G8_T4_S113_L00</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>448 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               SampleID  PL01  PL02  PL03  PL04  PL05  PL05a  PL05b\n",
       "0       2H9_T44_S98_L00     0     0     0     0     0      0      0\n",
       "1        2H9_T4_S96_L00     0     0     0     0     0      0      0\n",
       "2     4026_T30_S103_L00     0     0     0     0     0      0      0\n",
       "3     4026_T44_S104_L00     0     0     0     0     0      0      0\n",
       "4      4026_T4_S102_L00     0     0     0     0     0      0      0\n",
       "..                  ...   ...   ...   ...   ...   ...    ...    ...\n",
       "443  ND5A5_T44_S112_L00     0     0     0     0     0      0      0\n",
       "444   ND5A5_T4_S110_L00     0     0     0     0     0      0      0\n",
       "445  ND6G8_T30_S114_L00     0     0     0     0     0      0      0\n",
       "446  ND6G8_T44_S115_L00     0     0     0     0     0      0      0\n",
       "447   ND6G8_T4_S113_L00     0     0     0     0     0      0      0\n",
       "\n",
       "[448 rows x 8 columns]"
      ]
     },
     "execution_count": 387,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = pd.read_csv(f'GENE2_NAME.txt', sep = '\\t')\n",
    "metadata.columns = ['SampleID', 'PlateID', 'Strain', 'Sampling Time', 'Sample Number', 'No Clue', 'Same']\n",
    "covariates = build_covariates(metadata)\n",
    "covariates.drop(columns = ['Strain', 'Sampling Time', 'Sample Number', 'No Clue', 'Same'], inplace = True)\n",
    "covariates    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 391,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'metasub' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-391-833373f069ea>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mj\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;36m3\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m      \u001b[1;31m#metasub = covariates[covariates['SampleID'].isin(timepoint.columns)].T\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 5\u001b[1;33m        \u001b[0mmetasub\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcolumns\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mmetasub\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      6\u001b[0m        \u001b[0mmetasub\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mstrip_names\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmetasub\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mT\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      7\u001b[0m        \u001b[0mmetasub\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34mf'{args.data_path}{timepoint_times[i]}batchcov.csv'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'metasub' is not defined"
     ]
    }
   ],
   "source": [
    " timepoint_times = ['T4_', 'T30_', 'T44_']\n",
    "i = 0\n",
    "for j in range(1,3):\n",
    "      #metasub = covariates[covariates['SampleID'].isin(timepoint.columns)].T\n",
    "        metasub.columns = metasub.iloc[0,:]\n",
    "        metasub = strip_names(metasub).T\n",
    "        metasub.to_csv(f'{args.data_path}{timepoint_times[i]}batchcov.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
