{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import argparse\n",
    "\n",
    "\n",
    "def fix_names(countdata):\n",
    "    '''\n",
    "    This fixes the name specific issues in the NF54GFPxNHP4026 cross.\n",
    "    The names have changed several times and been recorded in different\n",
    "    formats, so I'll fix them with this.\n",
    "    '''\n",
    "    countdata = countdata.T\n",
    "    countdata.index = countdata.index.str.replace('\\/','', regex = True)\n",
    "    countdata.index = countdata.index.str.replace('ND5A5', 'AC075', regex = True)\n",
    "    countdata.index = countdata.index.str.replace('ND6G8', 'AC125', regex = True)\n",
    "    countdata.index = countdata.index.str.replace('N1', '', regex = True)\n",
    "    countdata.index = countdata.index.str.replace('\\\\.', '', regex = True)\n",
    "    countdata.index = countdata.index.str.replace('_4026', '_NHP4026', regex = True)\n",
    "    countdata.index = countdata.index.str.replace('^4026', 'NHP4026', regex = True)\n",
    "    countdata.index = countdata.index.str.replace('2H9', 'AC030', regex = True)\n",
    "    countdata.index = countdata.index.str.replace('6E5', 'AC033', regex = True)\n",
    "    return countdata.T\n",
    "   \n",
    "\n",
    "\n",
    "def pick_best_reps(countdata):\n",
    "    '''\n",
    "    Get this, the GTEx Consortium handles TECHNICAL replicates\n",
    "    (please don't mistake these for BIOLOGICAL replicates, which\n",
    "    are amazing and useful) by simply choosing the replicate\n",
    "    with the most reads. Good enough for the Broad, good enough\n",
    "    for me, so I'll do that here. Note that this contains a\n",
    "    pile of regular expressions for name handling specific to the\n",
    "    formats used in this cross. \n",
    "    '''\n",
    "    \n",
    "    # Calculate sample count sums\n",
    "    countdata = countdata.T\n",
    "    good_samples = []\n",
    "    sample_sum = []\n",
    "    for i in range(6,len(countdata.index)):\n",
    "        genes_rep = np.sum(countdata.iloc[i,:] > 0)\n",
    "        good_samples.append(countdata.index[i])\n",
    "        sample_sum.append(np.sum(countdata.iloc[i,:]))\n",
    "        \n",
    "    # Build summary frame, iterate over samples by strain/hpi and choose \n",
    "    # the sample with the most reads to keep\n",
    "\n",
    "    count_summary = {'Sample Name':good_samples,\n",
    "                     'Total Counts':sample_sum}\n",
    "    count_summary = pd.DataFrame(count_summary)\n",
    "\n",
    "    # This is a hot mess of regular expression that just converts the entire\n",
    "    # sample name to strain_##, where ## is the sampling timepoint. The formatting\n",
    "    # is very inconsistent throughout, so a mess of replacements need to be carefully\n",
    "    # made. This mess of substitutions makes them.\n",
    "\n",
    "    count_summary['Strain and Time'] = count_summary['Sample Name'].str.replace('^GF_PL[\\d]+[a,b,c]{0,1}_', '',regex = True)\n",
    "    count_summary['Strain and Time'] = count_summary['Strain and Time'].str.replace('[A,B,C]_', '', regex = True)\n",
    "    count_summary['Strain and Time'] = count_summary['Strain and Time'].str.replace('_[d]+$', '', regex = True)\n",
    "    count_summary['Strain and Time'] = count_summary['Strain and Time'].str.replace('_S.*', '', regex = True)\n",
    "    count_summary['Strain and Time'] = count_summary['Strain and Time'].str.replace('_[0-9]{3,4}$', '', regex = True)\n",
    "    count_summary['Strain and Time'] = count_summary['Strain and Time'].str.replace('hpi', '', regex = True)\n",
    "    count_summary['Strain and Time'] = count_summary['Strain and Time'].str.replace('_T', '_', regex = True)\n",
    "\n",
    "    # Believe it or not, the GTEx consortium handled reps across batches and such by...\n",
    "    # taking the replicate with the most counts. That's easy, let's do that.\n",
    "\n",
    "    best_samples = []\n",
    "    for sample in count_summary['Strain and Time'].unique():\n",
    "        subframe = count_summary[count_summary['Strain and Time'] == sample]\n",
    "        best_samples.append(count_summary.iloc[subframe['Total Counts'].idxmax(),:])\n",
    "    \n",
    "    # Now I'll build a frame with just the best samples in it\n",
    "    \n",
    "    best_samples = pd.DataFrame(best_samples)\n",
    "    best_samplenames = ['Geneid', 'Chr', 'Start', 'End', 'Strand', 'Length']\n",
    "    best_samplenames.extend(list(best_samples['Sample Name']))\n",
    "    curated_counts = countdata[countdata.index.isin(best_samplenames)]\n",
    "    return curated_counts.T\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def split_times(rawcounts):\n",
    "    '''\n",
    "    In this particular study I took 3 time points; 4hpi, 30hpi,\n",
    "    and 44hpi. This function takes a full set of count data and\n",
    "    splits it up in to 3 DataFrames. Note that it's hardcoded for the\n",
    "    time points I took, so it's not suitable for all crosses.\n",
    "    '''\n",
    "\n",
    "    T4_samples = rawcounts.filter(regex='T4_|_4hpi')\n",
    "    T4_samples = rawcounts.iloc[:,0:5].merge(T4_samples, \n",
    "                                             left_index = True, \n",
    "                                             right_index = True)\n",
    "    T30_samples = rawcounts.filter(regex='T30_|_30hpi')\n",
    "    T30_samples = rawcounts.iloc[:,0:5].merge(T30_samples, \n",
    "                                             left_index = True, \n",
    "                                             right_index = True)\n",
    "    T44_samples = rawcounts.filter(regex='T44_|_44hpi')\n",
    "    T44_samples = rawcounts.iloc[:,0:5].merge(T44_samples, \n",
    "                                             left_index = True, \n",
    "                                             right_index = True)\n",
    "\n",
    "\n",
    "\n",
    "    # Let's pull out the count data for each time point\n",
    "    # Now that we're normalized we can minimize this to a pure count matrix\n",
    "\n",
    "    return [T4_samples, T30_samples, T44_samples]\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "def build_covariates(metadata):\n",
    "    '''\n",
    "    Our samples were run across different plates; this\n",
    "    is the only real place I can account for technical\n",
    "    variation. Our sampling batches are confounded with\n",
    "    stage, so I am forced to rely on CRC and PEER to\n",
    "    pull anything in there out. To run PEER, I need\n",
    "    one-hot encoded plate data, so I do that here.\n",
    "    '''\n",
    "    encoding = pd.get_dummies(metadata['PlateID'])\n",
    "    metadata = metadata.merge(encoding,\n",
    "                              left_index = True, \n",
    "                              right_index = True)\n",
    "    metadata.drop(columns = 'PlateID',\n",
    "                 inplace = True)\n",
    "    \n",
    "    return metadata\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def fix_vcf(vcf_file, progenydata, vcf_out):\n",
    "    '''\n",
    "    The vcf file contains TX versions of strain names; this\n",
    "    function swaps them with the correct version.\n",
    "    '''\n",
    "    # Create dict for progeny data\n",
    "    \n",
    "    progenydata.index = progenydata.iloc[:,1]\n",
    "    progenydict = progenydata.iloc[:,0]\n",
    "    progenydict = progenydict.to_dict()\n",
    "    \n",
    "    \n",
    "    \n",
    "    vcf = open(vcf_file, 'r')\n",
    "    vcfout = open(vcf_out, 'w')\n",
    "    for line in vcf:\n",
    "        for key in progenydict.keys():\n",
    "            line = re.sub(key, progenydict[key], line)\n",
    "        vcfout.write(line)\n",
    "    \n",
    "    vcf.close()\n",
    "    vcfout.close()\n",
    "\n",
    "def strip_names(countdata):\n",
    "    '''\n",
    "    This function strips the long sample names from the\n",
    "    count data down to their strain only. NOTE: You really\n",
    "    don't want to use this until after you've built the metadata\n",
    "    file, as you need the full sample names to get the correct\n",
    "    Plate Number for batch correction.\n",
    "    '''\n",
    "    \n",
    "    countdata.columns = countdata.columns.str.replace('^GF[\\d]*_', '', regex = True)\n",
    "    countdata.columns = countdata.columns.str.replace('PL[\\da-z]*_', '', regex = True)\n",
    "    countdata.columns = countdata.columns.str.replace('[A-C]{1}_', '', regex = True)\n",
    "    countdata.columns = countdata.columns.str.partition('_').to_frame().iloc[:,0]\n",
    "    \n",
    "    return countdata\n",
    "\n",
    "def to_gct(file, expdata):\n",
    "\trow = len(expdata.index)\n",
    "\tcol = len(expdata.columns)\n",
    "\toutfile = open(file, 'w')\n",
    "\toutfile.write(f'1.0\\n')\n",
    "\toutfile.write(f'{row}\\t{col}\\n')\n",
    "\toutfile.close()\n",
    "\texpdata.to_csv(file,\n",
    "                       mode = 'a',\n",
    "                       header = True,\n",
    "                       sep = '\\t',\n",
    "                       encoding='utf-8',\n",
    "                       index = False)\n",
    "\n",
    "########################################################\n",
    "#\n",
    "# Main block- parses all files in one pass, if you're\n",
    "# feeling lazy\n",
    "#\n",
    "########################################################\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "counts = pd.read_csv(f'GENE2.count', sep = '\\t')\n",
    "counts = fix_names(counts)\n",
    "counts = pick_best_reps(counts)\n",
    "timepoints = split_times(counts)[0]\n",
    "probeinfo = counts.iloc[:,0:6]\n",
    "probeinfo.to_csv(f'test_probeinfo.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PL01</th>\n",
       "      <th>PL02</th>\n",
       "      <th>PL03</th>\n",
       "      <th>PL04</th>\n",
       "      <th>PL05</th>\n",
       "      <th>PL05a</th>\n",
       "      <th>PL05b</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AC030_T30_S97_L00</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC030_T44_S98_L00</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC030_T4_S96_L00</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NHP4026_T30_S103_L00</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NHP4026_T44_S104_L00</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC075_T44_S112_L00</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC075_T4_S110_L00</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC125_T30_S114_L00</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC125_T44_S115_L00</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC125_T4_S113_L00</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>449 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      PL01  PL02  PL03  PL04  PL05  PL05a  PL05b\n",
       "AC030_T30_S97_L00        0     0     0     0     0      0      0\n",
       "AC030_T44_S98_L00        0     0     0     0     0      0      0\n",
       "AC030_T4_S96_L00         0     0     0     0     0      0      0\n",
       "NHP4026_T30_S103_L00     0     0     0     0     0      0      0\n",
       "NHP4026_T44_S104_L00     0     0     0     0     0      0      0\n",
       "...                    ...   ...   ...   ...   ...    ...    ...\n",
       "AC075_T44_S112_L00       0     0     0     0     0      0      0\n",
       "AC075_T4_S110_L00        0     0     0     0     0      0      0\n",
       "AC125_T30_S114_L00       0     0     0     0     0      0      0\n",
       "AC125_T44_S115_L00       0     0     0     0     0      0      0\n",
       "AC125_T4_S113_L00        0     0     0     0     0      0      0\n",
       "\n",
       "[449 rows x 7 columns]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = pd.read_csv(f'GENE2_NAME.txt', sep = '\\t', header = None, index_col = 0)\n",
    "metadata.columns = ['PlateID', 'Strain', 'Sampling Time', 'Sample Number', 'No Clue', 'Same']\n",
    "covariates = build_covariates(metadata)\n",
    "covariates.drop(columns = ['Strain', 'Sampling Time', 'Sample Number', 'No Clue', 'Same'], inplace = True)\n",
    "covariates.rename_axis(None, inplace = True)\n",
    "covariates = fix_names(covariates.T).T\n",
    "covariates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PL01</th>\n",
       "      <th>PL02</th>\n",
       "      <th>PL03</th>\n",
       "      <th>PL04</th>\n",
       "      <th>PL05</th>\n",
       "      <th>PL05a</th>\n",
       "      <th>PL05b</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AC033</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC004</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC006</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC008</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC025</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC027</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC028</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC034</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC049</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC050</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC056</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC088</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC094</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC096</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC098</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC100</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC105</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC110</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC120</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC124</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC125</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC127</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NHP4026</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC005</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC023</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC024</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC041</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC042</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC043</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC044</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC045</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC046</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC047</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC048</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC057</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC070</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC077</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC079</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC093</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC097</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC103</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC109</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC118</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC038</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC082</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC030</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC032</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC074</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC130</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NF54gfp</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC075</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         PL01  PL02  PL03  PL04  PL05  PL05a  PL05b\n",
       "0                                                  \n",
       "AC033       0     0     0     0     0      0      0\n",
       "AC004       0     1     0     0     0      0      0\n",
       "AC006       0     1     0     0     0      0      0\n",
       "AC007       0     1     0     0     0      0      0\n",
       "AC008       0     1     0     0     0      0      0\n",
       "AC025       0     1     0     0     0      0      0\n",
       "AC027       0     1     0     0     0      0      0\n",
       "AC028       0     1     0     0     0      0      0\n",
       "AC034       0     1     0     0     0      0      0\n",
       "AC049       0     1     0     0     0      0      0\n",
       "AC050       0     1     0     0     0      0      0\n",
       "AC056       0     1     0     0     0      0      0\n",
       "AC088       0     1     0     0     0      0      0\n",
       "AC094       0     1     0     0     0      0      0\n",
       "AC096       0     1     0     0     0      0      0\n",
       "AC098       0     1     0     0     0      0      0\n",
       "AC100       0     1     0     0     0      0      0\n",
       "AC105       0     1     0     0     0      0      0\n",
       "AC110       0     1     0     0     0      0      0\n",
       "AC120       0     1     0     0     0      0      0\n",
       "AC124       0     1     0     0     0      0      0\n",
       "AC125       0     1     0     0     0      0      0\n",
       "AC127       0     1     0     0     0      0      0\n",
       "NHP4026     0     1     0     0     0      0      0\n",
       "AC005       0     0     1     0     0      0      0\n",
       "AC023       0     0     1     0     0      0      0\n",
       "AC024       0     0     1     0     0      0      0\n",
       "AC041       0     0     1     0     0      0      0\n",
       "AC042       0     0     1     0     0      0      0\n",
       "AC043       0     0     1     0     0      0      0\n",
       "AC044       0     0     1     0     0      0      0\n",
       "AC045       0     0     1     0     0      0      0\n",
       "AC046       0     0     1     0     0      0      0\n",
       "AC047       0     0     1     0     0      0      0\n",
       "AC048       0     0     1     0     0      0      0\n",
       "AC057       0     0     1     0     0      0      0\n",
       "AC070       0     0     1     0     0      0      0\n",
       "AC077       0     0     1     0     0      0      0\n",
       "AC079       0     0     1     0     0      0      0\n",
       "AC093       0     0     1     0     0      0      0\n",
       "AC097       0     0     1     0     0      0      0\n",
       "AC103       0     0     1     0     0      0      0\n",
       "AC109       0     0     1     0     0      0      0\n",
       "AC118       0     0     1     0     0      0      0\n",
       "AC038       0     0     0     0     0      1      0\n",
       "AC082       0     0     0     0     0      1      0\n",
       "AC030       0     0     0     0     0      0      1\n",
       "AC032       0     0     0     0     0      0      1\n",
       "AC074       0     0     0     0     0      0      1\n",
       "AC130       0     0     0     0     0      0      1\n",
       "NF54gfp     0     0     0     0     0      0      1\n",
       "AC075       0     0     0     0     0      0      0"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metasub = metadata\n",
    "timepoint = timepoints\n",
    "metasub = covariates[covariates.index.isin(timepoint.columns)].T\n",
    "metasub = strip_names(metasub).T\n",
    "# NOTE: we have RNAseq for this parasite, but no genome seq\n",
    "metasub.drop(index = 'AC081', inplace = True)\n",
    "#metasub.to_csv(f'{args.data_path}{timepoint_times[i]}batchcov.csv')\n",
    "timepoint.drop(columns = ['Start', 'End', 'Strand', 'Chr'], inplace = True)\n",
    "timepoint = strip_names(timepoint)\n",
    "timepoint.drop(columns = 'AC081', inplace = True)\n",
    "#timepoint.to_csv(f'{args.data_path}{timepoint_times[i]}counts.txt',\n",
    "#        sep = '\\t',\n",
    "#        index = False)  \n",
    "metasub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pfal_genomeandanno_prep as pfal\n",
    "import os\n",
    "import argparse\n",
    "import numpy as np\n",
    "\n",
    "def sample_and_geneqc(expdata):\n",
    "    '''\n",
    "    Here, we'll remove poor samples and poor genes, and QC low read #s\n",
    "    Low reads are any gene/sample reading < 5 counts; those are zeroed out\n",
    "    Poor samples are defined as those with < 3000 genes with reads\n",
    "    Poor genes are those that appear in < 20% of samples\n",
    "    '''\n",
    "    \n",
    "    # Curate Samples\n",
    "    \n",
    "    expdata = expdata.mask(expdata < 5, 0)\n",
    "    \n",
    "    genecounts = pd.Series(data = np.count_nonzero(expdata, axis = 1),\n",
    "                             index = expdata.index)\n",
    "    samplecounts = pd.Series(np.count_nonzero(expdata,axis = 0),\n",
    "                          index = expdata.columns)\n",
    "    \n",
    "    goodgenes = genecounts[genecounts/samplecounts.size > 0.2]\n",
    "    \n",
    "    goodsamples = samplecounts[samplecounts > 3000]\n",
    "    \n",
    "    allcur = expdata.loc[goodgenes.index, goodsamples.index]\n",
    "    \n",
    "    # Zero out counts < 5\n",
    "    \n",
    "    \n",
    "    \n",
    "    return allcur\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AC033</th>\n",
       "      <th>AC004</th>\n",
       "      <th>AC006</th>\n",
       "      <th>AC008</th>\n",
       "      <th>AC027</th>\n",
       "      <th>AC028</th>\n",
       "      <th>AC034</th>\n",
       "      <th>AC049</th>\n",
       "      <th>AC050</th>\n",
       "      <th>AC056</th>\n",
       "      <th>...</th>\n",
       "      <th>AC103</th>\n",
       "      <th>AC109</th>\n",
       "      <th>AC118</th>\n",
       "      <th>AC038</th>\n",
       "      <th>AC082</th>\n",
       "      <th>AC030</th>\n",
       "      <th>AC032</th>\n",
       "      <th>AC074</th>\n",
       "      <th>AC130</th>\n",
       "      <th>NF54gfp</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>malmito_rna_LSUG:rRNA</th>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>malmito_SSUB:rRNA</th>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>malmito_rna_1:rRNA</th>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>24</td>\n",
       "      <td>7</td>\n",
       "      <td>18</td>\n",
       "      <td>9</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>22</td>\n",
       "      <td>25</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>29</td>\n",
       "      <td>88</td>\n",
       "      <td>21</td>\n",
       "      <td>16</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>malmito_rna_10:rRNA</th>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>38</td>\n",
       "      <td>14</td>\n",
       "      <td>26</td>\n",
       "      <td>10</td>\n",
       "      <td>40</td>\n",
       "      <td>6</td>\n",
       "      <td>23</td>\n",
       "      <td>31</td>\n",
       "      <td>...</td>\n",
       "      <td>29</td>\n",
       "      <td>22</td>\n",
       "      <td>29</td>\n",
       "      <td>47</td>\n",
       "      <td>48</td>\n",
       "      <td>38</td>\n",
       "      <td>130</td>\n",
       "      <td>31</td>\n",
       "      <td>29</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mal_mito_1</th>\n",
       "      <td>18</td>\n",
       "      <td>11</td>\n",
       "      <td>21</td>\n",
       "      <td>12</td>\n",
       "      <td>16</td>\n",
       "      <td>11</td>\n",
       "      <td>25</td>\n",
       "      <td>14</td>\n",
       "      <td>23</td>\n",
       "      <td>18</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>11</td>\n",
       "      <td>40</td>\n",
       "      <td>13</td>\n",
       "      <td>30</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF3D7_1478600</th>\n",
       "      <td>11</td>\n",
       "      <td>15</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>32</td>\n",
       "      <td>24</td>\n",
       "      <td>28</td>\n",
       "      <td>19</td>\n",
       "      <td>9</td>\n",
       "      <td>15</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>23</td>\n",
       "      <td>10</td>\n",
       "      <td>26</td>\n",
       "      <td>14</td>\n",
       "      <td>29</td>\n",
       "      <td>28</td>\n",
       "      <td>47</td>\n",
       "      <td>14</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF3D7_1478800</th>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>44</td>\n",
       "      <td>37</td>\n",
       "      <td>103</td>\n",
       "      <td>118</td>\n",
       "      <td>121</td>\n",
       "      <td>50</td>\n",
       "      <td>64</td>\n",
       "      <td>156</td>\n",
       "      <td>...</td>\n",
       "      <td>69</td>\n",
       "      <td>59</td>\n",
       "      <td>43</td>\n",
       "      <td>49</td>\n",
       "      <td>56</td>\n",
       "      <td>95</td>\n",
       "      <td>116</td>\n",
       "      <td>74</td>\n",
       "      <td>72</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF3D7_1478900</th>\n",
       "      <td>127</td>\n",
       "      <td>209</td>\n",
       "      <td>199</td>\n",
       "      <td>101</td>\n",
       "      <td>318</td>\n",
       "      <td>107</td>\n",
       "      <td>563</td>\n",
       "      <td>168</td>\n",
       "      <td>404</td>\n",
       "      <td>131</td>\n",
       "      <td>...</td>\n",
       "      <td>190</td>\n",
       "      <td>248</td>\n",
       "      <td>177</td>\n",
       "      <td>867</td>\n",
       "      <td>345</td>\n",
       "      <td>231</td>\n",
       "      <td>170</td>\n",
       "      <td>232</td>\n",
       "      <td>96</td>\n",
       "      <td>466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF3D7_1479000</th>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>74</td>\n",
       "      <td>7</td>\n",
       "      <td>18</td>\n",
       "      <td>287</td>\n",
       "      <td>111</td>\n",
       "      <td>158</td>\n",
       "      <td>14</td>\n",
       "      <td>...</td>\n",
       "      <td>73</td>\n",
       "      <td>217</td>\n",
       "      <td>20</td>\n",
       "      <td>183</td>\n",
       "      <td>87</td>\n",
       "      <td>139</td>\n",
       "      <td>11</td>\n",
       "      <td>225</td>\n",
       "      <td>114</td>\n",
       "      <td>320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF3D7_1479200</th>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>33</td>\n",
       "      <td>49</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>39</td>\n",
       "      <td>27</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>34</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "      <td>22</td>\n",
       "      <td>24</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4802 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       AC033  AC004  AC006  AC008  AC027  AC028  AC034  AC049  \\\n",
       "Geneid                                                                          \n",
       "malmito_rna_LSUG:rRNA      0      7      5      0      0      0      6      0   \n",
       "malmito_SSUB:rRNA          0      5      0      0      0      0      0      0   \n",
       "malmito_rna_1:rRNA         0     23     24      7     18      9     16      0   \n",
       "malmito_rna_10:rRNA        0     23     38     14     26     10     40      6   \n",
       "mal_mito_1                18     11     21     12     16     11     25     14   \n",
       "...                      ...    ...    ...    ...    ...    ...    ...    ...   \n",
       "PF3D7_1478600             11     15     21      0     32     24     28     19   \n",
       "PF3D7_1478800              0     20     44     37    103    118    121     50   \n",
       "PF3D7_1478900            127    209    199    101    318    107    563    168   \n",
       "PF3D7_1479000              0     11     12     74      7     18    287    111   \n",
       "PF3D7_1479200              0      5      0     15     33     49      0     28   \n",
       "\n",
       "                       AC050  AC056  ...  AC103  AC109  AC118  AC038  AC082  \\\n",
       "Geneid                               ...                                      \n",
       "malmito_rna_LSUG:rRNA      0      5  ...      5      5     15      0      6   \n",
       "malmito_SSUB:rRNA          0      0  ...      0      8      5      0      0   \n",
       "malmito_rna_1:rRNA         7     13  ...     20     22     25     21     21   \n",
       "malmito_rna_10:rRNA       23     31  ...     29     22     29     47     48   \n",
       "mal_mito_1                23     18  ...     16     11     40     13     30   \n",
       "...                      ...    ...  ...    ...    ...    ...    ...    ...   \n",
       "PF3D7_1478600              9     15  ...     15     23     10     26     14   \n",
       "PF3D7_1478800             64    156  ...     69     59     43     49     56   \n",
       "PF3D7_1478900            404    131  ...    190    248    177    867    345   \n",
       "PF3D7_1479000            158     14  ...     73    217     20    183     87   \n",
       "PF3D7_1479200             29      0  ...     39     27      6      8     34   \n",
       "\n",
       "                       AC030  AC032  AC074  AC130  NF54gfp  \n",
       "Geneid                                                      \n",
       "malmito_rna_LSUG:rRNA      5     12      0      0       10  \n",
       "malmito_SSUB:rRNA          0      9      0      0        0  \n",
       "malmito_rna_1:rRNA        29     88     21     16       26  \n",
       "malmito_rna_10:rRNA       38    130     31     29       44  \n",
       "mal_mito_1                19     10      8      5       30  \n",
       "...                      ...    ...    ...    ...      ...  \n",
       "PF3D7_1478600             29     28     47     14       22  \n",
       "PF3D7_1478800             95    116     74     72       69  \n",
       "PF3D7_1478900            231    170    232     96      466  \n",
       "PF3D7_1479000            139     11    225    114      320  \n",
       "PF3D7_1479200              5      9     22     24       14  \n",
       "\n",
       "[4802 rows x 48 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expdata = pd.read_csv('T4_counts.txt', sep = '\\t', index_col = 0)\n",
    "test = sample_and_geneqc(expdata)\n",
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
